lib/perl/Genome/Model/Build/ReferenceSequence/IndexBase.pm

```diff
@@ -0,0 +1,277 @@
+        reference_build => {
+            is => 'Genome::Model::Build::ImportedReferenceSequence',
+            id_by => 'reference_build_id',
+        },
+        reference_name => {
+            via => 'reference_build',
+            to => 'name',
+            is_mutable => 0,
+            is_optional => 1
+        },
+        aligner => {
+            calculate_from => [qw/aligner_name aligner_version aligner_params/],
+            calculate => q|no warnings; "$aligner_name $aligner_version $aligner_params"|
+        },
+    ],
+    has_input => [
+        reference_build_id  => {
+            is => 'Number',
+            doc => 'the reference to use by id',
+        },
+    ],
+    has_param => [
+        aligner_name => {
+            is => 'Text', default_value => 'maq',
+            doc => 'the name of the aligner to use, maq, blat, newbler etc.',
+        },
+        aligner_version => {
+            is => 'Text',
+            doc => 'the version of the aligner to use, i.e. 0.6.8, 0.7.1, etc.',
+            is_optional=>1,
+        },
+        aligner_params => {
+            is => 'Text',
+            is_optional=>1,
+            doc => 'any additional params for the aligner in a single string',
+        },
+    ],
+    has_transient => [
+        aligner_class_name => {
+            is => 'Text',
+            is_optional => 1,
+        }
+    ]
+};
+
+sub create {
+    my $class = shift;
+    my %p = @_;
+
+    my $aligner_class = 'Genome::InstrumentData::AlignmentResult::'  . Genome::InstrumentData::AlignmentResult->_resolve_subclass_name_for_aligner_name($p{aligner_name});
+    $class->debug_message(sprintf("Resolved aligner class %s, making sure it's real and can be loaded.", $aligner_class));
+    unless ($aligner_class->class) {
+        $class->error_message(sprintf("Failed to load aligner class (%s).", $aligner_class));
+        return;
+    }
+
+    my $self = $class->SUPER::create(%p);
+    return unless $self;
+    $self->aligner_class_name($aligner_class);
+
+    $self->debug_message("Prepare staging directories...");
+    unless ($self->_prepare_staging_directory) {
+        $self->error_message("Failed to prepare working directory");
+        return;
+    }
+
+    my $reference_fasta_file = $self->_resolve_reference_fasta_file;
+    unless (-s $reference_fasta_file) {
+        $self->error_message(sprintf("Reference fasta file %s does not exist", $reference_fasta_file));
+        return;
+    } else {
+        $self->debug_message(sprintf("Confirmed non-zero reference fasta file is %s", $reference_fasta_file));
+    }
+
+    unless ($self->_prepare_index($reference_fasta_file)) {
+        $self->error_message("Failed to prepare reference index!");
+        return;
+    }
+
+    unless ($self->_promote_prepared_index) {
+        $self->error_message("Failed to save reference index to output location!");
+        return;
+    }
+
+    unless ($self->generate_dependencies_as_needed($self->_user_data_for_nested_results)) {
```

Comment:
Should this be added as an abstract method with a Carp::confess if it is not defined in the subclass?